cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -0,0 +1,101 @@
+"""
+Resolve the dependency list for a yarn project.
+
+It also performs the necessary validations to avoid allowing an invalid
+project to keep being processed.
+"""
+import logging
+from typing import NamedTuple
+
+from cachi2.core.models.sbom import Component
+from cachi2.core.package_managers.yarn.locators import Locator
+from cachi2.core.package_managers.yarn.project import YarnRc
+from cachi2.core.package_managers.yarn.utils import run_yarn_cmd
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(__name__)
+
+
+class Cache(NamedTuple):
+    """Cache information for a package."""
+
+    checksum: str
+    path: str
+
+
+class Package(NamedTuple):
+    """A package listed by the yarn info command.
+
+    See the output for 'yarn info -AR --json --cache'.
+
+    Note that the attributes present in the json vary a little depending on the
+    type of dependency.
+    """
+
+    cache: Cache
+    locator: Locator
+    version: str
+
+    @classmethod
+    def from_info_string(cls, info: str) -> "Package":
+        """Create a Package from the output of yarn info."""
+        # this should use the locators.parse_locator function
+        return NotImplemented
+
+
+def resolve_dependency_info(source_dir: RootedPath) -> list[Package]:
+    """Fetch and parse dependency info from the 'yarn info' output.
+
+    This function also performs a validation to ensure that the current
+    yarn project can be processed.
+
+    :raises PackageRejected: if the validation fails.
+    :raises SubprocessCallError: if the 'yarn info' command fails.
+    """
+    result = run_yarn_cmd([], source_dir)
+
+    # the result is not a valid json list, but a sequence of json objects separated by line breaks
+    packages = [Package.from_info_string(info) for info in result.splitlines()]
+
+    _vet_git_dependencies(packages)
+
+    return packages
+
+
+def create_component_from_package(package: Package, yarn_rc: YarnRc) -> Component:
+    """Create a SBOM component from a yarn Package."""
+    name = _resolve_package_name(package)
+
+    return Component(
+        name=name, version=package.version, purl=_generate_purl_for_package(package, name, yarn_rc)
+    )
```

Comment:
Just remembered we're forgetting the `pedigree.patches` for patch: dependencies, but that will require changes to the Component model as well

And from the code "interface" side of things, that probably just means we need to pass `Project` or at least its `source_dir` to this function so that it can handle file paths (in fact, I think we need that for all file-like dependencies)

Reply:
I added the project as a parameter to this function, and to the _generate_purl_for_package function (we'd probably need to split the latter into several functions, but I'll leave this to the implementation).